### ISSUE TITLE
Error doesn't decrease

### ISSUE THREAD

**Issue by shrubb** [freesouls/face-alignment-at-3000fps#4]
Hi,

First of all, thanks for this implementation.
I ran and trained successfully the previous version (before refactoring), but now I have troubles with training. As before, I followed the instructions in README, downloaded HELEN dataset, put it in `example/helen`. The reported error rate is:

```
training stage: 0 of 6
train regression error: 1071.14, mean error: 0.149516
Validation at stage: 0
Validation error: 37.3754, mean error: 0.120178

training stage: 1 of 6
train regression error: 847.716, mean error: 0.11833
Validation at stage: 1
Validation error: 32.5232, mean error: 0.104576

training stage: 2 of 6
train regression error: 735.792, mean error: 0.102707
Validation at stage: 2
Validation error: 30.7561, mean error: 0.0988942

training stage: 3 of 6
train regression error: 803.279, mean error: 0.112127
Validation at stage: 3
Validation error: 36.0491, mean error: 0.115914

training stage: 4 of 6
train regression error: 752.986, mean error: 0.105107
Validation at stage: 4
Validation error: 35.4158, mean error: 0.113877

training stage: 5 of 6
train regression error: 791.054, mean error: 0.110421
Validation at stage: 5
Validation error: 38.1961, mean error: 0.122817

finish training, start to saving the model...
model name: helenModel
save the model successfully
```

Obviously, the quality is then rather poor:

![landmarks](https://cloud.githubusercontent.com/assets/9570420/13630707/c0668b3e-e5f1-11e5-8f51-116ba850ef80.png)

Maybe I'm doing something wrong?

**Comment by freesouls** [comment]
please wait, let me verify this!

**Comment by shrubb** [comment]
@freesouls 
Okay, thank you!

**Comment by freesouls** [comment]
Hi, I just clone the codes to a new folder, and run the program(without changing anything), the output is below, I just run 3 stages, it seems the error is decreasing (thought _different from_ the output in `README.md`, because `initial_guess` is **random** sampled, `positions` of each pixel difference features are also **random** sampled, the `threshold` when spliting nodes in random forest is also **random** selected when you run the program)
```
training stage: 0 of 6
train regression error: 897.343, mean error: 0.125257
Validation at stage: 0
Validation error: 31.7507, mean error: 0.102092

training stage: 1 of 6
train regression error: 629.442, mean error: 0.0878617
Validation at stage: 1
Validation error: 25.116, mean error: 0.0807589

training stage: 2 of 6
train regression error: 493.326, mean error: 0.0688618
Validation at stage: 2
Validation error: 22.7215, mean error: 0.0730595
```
will you rerun the codes?

**Comment by shrubb** [comment]
Hm, let me try this on a different machine in a moment. That run was on a 24-core machine, maybe this caused the problem.

**Comment by freesouls** [comment]
My machine is 8-core(the program is running in 8 threads at the same time), but the codes should be OK with 24 threads. Off course, may be there are something wrong with the codes.

**Comment by freesouls** [comment]
Hi, before running, set `params.local_features_num_ = 400 or 500`, 200 may be too small, and it won't effect the running time, for linear regression takes about 80% training time

**Comment by shrubb** [comment]
Thank you for helping. I've set local features number to 500. That's very strange: I got it working on a desktop machine, but no effect on a remote one...
I'll continue investigating this!

**Comment by freesouls** [comment]
sometimes, for the same image, opencv's(even using the same version) face detector will give different results.

**Comment by xiaoxiongli** [comment]
Dear freesouls:
        i try to train a model using your default settings, the training loss seems big comparing your training log:
train regression error: 575.186, mean error: **0.0799759**
Validation at stage: 5
Validation error: 28.2551, mean error: **0.0905614** 

and after the model is generated, i test one picture, it seems the result is not good:

![image](https://cloud.githubusercontent.com/assets/13975044/21496047/97bc7238-cc56-11e6-9daf-f2c84e0db4ed.png)

the whole training log:

use [./application train train_config_file] to train models
    [./application test test_config_file] to test images


parsing config_file: ../example/helen_train_config.txt

model name is: helenModel
local_features_num_: 500
landmarks_num_per_face_: 68
regressor_stages_: 6
tree_depth_: 5
trees_num_per_forest_: 12
overlap_: 0.3
initial_guess_: 3
local_radius_by_stages_: 0.29 0.21 0.16 0.12 0.08 0.04

reading training images paths:
../example/helen/trainset/
../example/helen_train_images_list.txt

reading validation images paths:
../example/helen/testset/
../example/helen_test_images_list_with_ground_truth.txt
parsing config file done

Loading training dataset...
face detector loaded : 1
loading images...
loading images in folder: ../example/helen/trainset/
100 images loaded
200 images loaded
300 images loaded
400 images loaded
500 images loaded
600 images loaded
700 images loaded
800 images loaded
900 images loaded
1000 images loaded
1100 images loaded
1200 images loaded
1300 images loaded
1400 images loaded
1500 images loaded
1600 images loaded
1700 images loaded
get 1798 faces in ../example/helen/trainset/
get 1798 faces in total

Loading validation dataset...
face detector loaded : 1
loading images...
loading images in folder: ../example/helen/testset/
100 images loaded
200 images loaded
300 images loaded
get 312 faces in ../example/helen/testset/
get 312 faces in total
Start training...
augment data sets
augmented size: 7192
training stage: 0 of 6
calculate regression targets
train forest of stage:0
it will take some time to build the Random Forest, please be patient!!!
Get Global Binary Features

Global Regression of stage 0
it will take some time to do Linear Regression, please be patient!!!
regressing ...0
regressing ...24
regressing ...64
regressing ...48
regressing ...16
regressing ...40
regressing ...8
regressing ...56
regressing ...32
predict regression targets
update current shapes
train regression error: 1056.25, mean error: 0.146864
Validation at stage: 0
Validation error: 36.3435, mean error: 0.116486
training stage: 1 of 6
calculate regression targets
train forest of stage:1
it will take some time to build the Random Forest, please be patient!!!
Get Global Binary Features

Global Regression of stage 1
it will take some time to do Linear Regression, please be patient!!!
regressing ...regressing ...0
48
regressing ...regressing ...6424

regressing ...40
regressing ...16
regressing ...8
regressing ...32
regressing ...56
predict regression targets
update current shapes
train regression error: 867.414, mean error: 0.120608
Validation at stage: 1
Validation error: 32.9086, mean error: 0.105476
training stage: 2 of 6
calculate regression targets
train forest of stage:2
it will take some time to build the Random Forest, please be patient!!!
Get Global Binary Features

Global Regression of stage 2
it will take some time to do Linear Regression, please be patient!!!
regressing ...regressing ...0
regressing ...48
regressing ...24
64
regressing ...40
regressing ...16
regressing ...32
regressing ...8
regressing ...56
predict regression targets
update current shapes
train regression error: 757.218, mean error: 0.105286
Validation at stage: 2
Validation error: 30.699, mean error: 0.0983942
training stage: 3 of 6
calculate regression targets
train forest of stage:3
it will take some time to build the Random Forest, please be patient!!!
Get Global Binary Features

Global Regression of stage 3
it will take some time to do Linear Regression, please be patient!!!
regressing ...regressing ...0
64
regressing ...24regressing ...
48
regressing ...40
regressing ...16
regressing ...8
regressing ...32
regressing ...56
predict regression targets
update current shapes
train regression error: 691.209, mean error: 0.0961081
Validation at stage: 3
Validation error: 29.9347, mean error: 0.0959444
training stage: 4 of 6
calculate regression targets
train forest of stage:4
it will take some time to build the Random Forest, please be patient!!!
Get Global Binary Features

Global Regression of stage 4
it will take some time to do Linear Regression, please be patient!!!
regressing ...regressing ...0
48
regressing ...64regressing ...
24
regressing ...40
regressing ...16
regressing ...8
regressing ...32
regressing ...56
predict regression targets
update current shapes
train regression error: 614.908, mean error: 0.0854989
Validation at stage: 4
Validation error: 28.6554, mean error: 0.0918444
training stage: 5 of 6
calculate regression targets
train forest of stage:5
it will take some time to build the Random Forest, please be patient!!!
Get Global Binary Features

Global Regression of stage 5
it will take some time to do Linear Regression, please be patient!!!
regressing ...0
regressing ...64regressing ...
24
regressing ...48
regressing ...40
regressing ...16
regressing ...8
regressing ...32
regressing ...56
predict regression targets
update current shapes
train regression error: 575.186, mean error: 0.0799759
Validation at stage: 5
Validation error: 28.2551, mean error: 0.0905614
finish training, start to saving the model...
model name: helenModel
save the model successfully

**Comment by xiaoxiongli** [comment]
i think that there maybe some bug related with openmp or multi-thread, because i disable the openmp in the Cmakelist.txt,   things get better.......

**Comment by xiaoxiongli** [comment]
do not use multi-thread(openmp):

Global Regression of stage 5
it will take some time to do Linear Regression, please be patient!!!
regressing ...0
regressing ...8
regressing ...16
regressing ...24
regressing ...32
regressing ...40
regressing ...48
regressing ...56
regressing ...64
predict regression targets
update current shapes
train regression error: 177.771, mean error: 0.0247179
Validation at stage: 5
Validation error: 17.5888, mean error: 0.0563745
finish training, start to saving the model...
model name: helenModel
save the model successfully

![default](https://cloud.githubusercontent.com/assets/13975044/21512611/509923f8-ccea-11e6-8566-5440ea2e39e5.jpg)
